### DISCUSSION TITLE
Syntax to highlight the partial-ed callable

### DISCUSSION THREAD

**Discussion by hsfzxjy** [abersheeran/cool#6] [Ideas]
Hi. I was thinking that the syntax `F(f, arg)` is sometimes less readable, since at the first sight it’s hard to tell “what is the callable” from the expression, especially if the expression is long, or `f` is not a mere identifier. 

I was curious about if you are interested in providing some alternative syntax like following:
```python
f @ F(arg1, arg2)  # equiv to F(f, arg1, arg2)
f @ F                      # equiv to F(f)
```
I think it’s clearer to place callable at the very front of the expression. The operator `@` can be implemented using `__rmatmul__`. `@` has a higher precedence than `|` so it’s OK to write `arg1 | f @ F(..., arg2)`. Of course you can change `@` or `F` to anything you think is more suitable.

The `@` operator would be similar to `$` operator in Haskell to some extend.

**Comment by abersheeran** [comment]
That's a cool idea.
The current design is mainly to keep in line with the Python standard library. After PEP612 is officially used, existing usage can get a lot of code hints from it.

**Comment by abersheeran** [comment]
I have invited you to be a member of this repository, https://github.com/abersheeran/cool/pull/7 and opened a new PR for this topic. If you are interested in implementing this call method, you can improve it (or close it and open a new PR), and then wait if more people approve this PR, we can merge it into the Master branch. Maybe this will waste a lot of time, but I will be very cautious about the addition of some features that are not particularly important. Just like #3.

**Comment by hsfzxjy** [comment]
Thank you very much for the quick implementation. I will check and improve it tomorrow.

I agree that these features need further discussion. “sugar”s should always be designed carefully.

**Comment by frostming** [comment]
As I suggested before, another option would be currying the `F`:

```python
args | F(map)(lambda x: x ** 2)
```

For the implementation details, `F.__call__` should return a new `F` object with arguments supplied. This approach benefits in the following ways:

1. One can store a partially initialized `F` in a variable: `Map = F(map)`
2. This is compatible with the current usage: `F(map)(arg) == F(map, arg)`

**Reply by abersheeran** [reply]
`D = lambda func: F(F, func)` can implement your ideas.

```python
range(10) | D(map)(lambda x: x)
```

**Reply by frostming** [reply]
Cool
